helper_thread.c

```diff
@@ -188,14 +193,36 @@ static uint8_t wait_for_action(int fd, unsigned int timeout_ms)
 	FD_SET(fd, &rfds);
 	FD_ZERO(&efds);
 	FD_SET(fd, &efds);
-	res = select(fd + 1, &rfds, NULL, &efds, &timeout);
+#ifdef CONFIG_HAVE_TIMERFD_CREATE
+	{
+		/*
+		 * If the timer frequency is 100 Hz, select() will round up
+		 * `timeout` to the next multiple of 1 / 100 Hz = 10 ms. Hence
+		 * use a high-resolution timer if possible to increase
+		 * select() timeout accuracy.
+		 */
+		struct itimerspec delta = {};
+
+		delta.it_value.tv_sec = timeout.tv_sec;
+		delta.it_value.tv_nsec = timeout.tv_usec * 1000;
+		res = timerfd_settime(timerfd, 0, &delta, NULL);
+		assert(res == 0);
+		FD_SET(timerfd, &rfds);
+	}
+#endif
+	res = select(max(fd, timerfd) + 1, &rfds, NULL, &efds,
+		     timerfd >= 0 ? NULL : &timeout);
 	if (res < 0) {
 		log_err("fio: select() call in helper thread failed: %s",
 			strerror(errno));
 		return A_EXIT;
 	}
 	if (FD_ISSET(fd, &rfds))
 		read_from_pipe(fd, &action, sizeof(action));
+	if (timerfd >= 0 && FD_ISSET(timerfd, &rfds)) {
+		res = read(timerfd, &exp, sizeof(exp));
+		assert(res == sizeof(exp));
```

Comment:
How frequently will the if body be called?

Reply:
Every time the timeout for `timerfd` has been reached. However, the time spent in the read() call should be small since all that call does is to read 8 bytes from kernel memory.